docs/dc/injected/models.py

```diff
@@ -0,0 +1,48 @@
+
+# models.py
+from dataclasses import dataclass
+
+from wired.dataclasses import factory, injected
+
+
+@dataclass
+class Request:
+    """ The app makes one of these and puts into the container """
+    url: str
+
+
+@dataclass
+class Customer:
+    """ A basic customer """
+    name: str
+
+
+CUSTOMERS = dict(
+    larry=Customer(name='Larry'),
+    anne=Customer(name='Anne')
+)
+
+
+@factory()
+@dataclass
+class Context:
+    """ Use request.url to retrieve a customer from the database """
+
+    request: Request
+
+    @classmethod
+    def wired_factory(cls, container):
+        request = container.get(Request)
+        customer = CUSTOMERS.get(request.url)
+        return customer
+
+
+@factory()
+@dataclass
+class Greeter:
+    """ A basic greeter """
+    customer: Customer = injected(Context)
```

Comment:
So the Context is acting as some form of proxy registered only to have a dynamic factory. An instance of Context is never actually returned by the `Context.wired_factory`. What can be done to make this more clear? Could we call `Context` a `DynamicCustomer` or something? Is a Context type actually necessary versus some sort of `injected(get_customer)`?

Reply:
This is very much a place I could use architectural help. I've changed my mind frequently and still think the abstraction is wrong.

Pyramid has Context and it has Resource. But not every Context is in the resource tree (e.g. NotFound). I was trying to avoid conflating Context and Resource.

From a static typing perspective, I'm not sure how to represent it. Some of the `injected()` machinery it to allow a view to know "this is going to return a Customer".

On your last question, `context` is such a first-class part of Pyramid that I thought it critical to always be able to get an instance.

Reply:
The issue I was pointing out in this example is that `greeter.customer` is never an instance of `Context`. Rather, the `Context` here is used only for its `wired_factory` which never returns an instance of `Context`. It's really a proxy that returns a type of customer based on the values in another service. It might be better to call it a `DynamicCustomer` and do something like `customer: Customer = injected(DynamicCustomer)`. Or you could take a callable, versus a type:

```
def get_dynamic_customer(container):
    request = container.get(Request)
    return CUSTOMERS[request.url]

customer: Customer = injected(factory=get_dynamic_customer)
```

Reply:
This is something I need to get right. I propose to move this conversation to a ticket where we can have a longer conversation. That ok?

Reply:
Yeah I think this probably needs to be discussed, the correct path forward is not super clear to me either.